FWCore/Framework/test/test_wrongOptionsType_cfg.py

```diff
@@ -0,0 +1,21 @@
+import FWCore.ParameterSet.Config as cms
+
+import argparse
+import sys
+
+parser = argparse.ArgumentParser(prog=sys.argv[0], description='Test wrong process.options parameter types')
+
+parser.add_argument("--name", help="Name of parameter", type=str)
+parser.add_argument("--value", help="Value of the parameter", type=str)
+
+argv = sys.argv[:]
+if '--' in argv:
+    argv.remove("--")
+args, unknown = parser.parse_known_args(argv)
+
+process = cms.Process("TEST")
+process.source = cms.Source("EmptySource")
+
+process.maxEvents.input = 2
+
+setattr(process.options, args.name, eval(str(args.value)))
```

Comment:
Should we consider making the setting function more recursive? I.e. if it is already set with an 'optional' it enforces the type checking? I'm guessing we didn't do that initially as some older parts wanted to be able to declare different types. But if we made 'optional' smarter to know it could have different types, maybe we could handle that case as well.

In that way optional could always catch type problems at the python level.

Reply:
I agree in being able to catch the problem already in python would be be better (at least for some patterns, I guess we can't do much for `del process.options.numberOfThreads; process.options.numberOfThreads = cms.untracked.int64(3)`).

A technical complication for stronger enforcement is that the assignment
```py
process.options.numberOfThreads = cms.untracked.int64(3)
```
executes code in the `process.options` PSet. So the `PSet.__setattr__()` would have to be made smarter (it looks doable though).

Another complication is that only one parameter in the  `process.options` PSet is `optional`, and the others (including `numberOfThreads`) have a default value set. So I think we'd need to think more how exactly to proceed.

All that sounds like a material for a new issue.

Reply:
I opened an issue https://github.com/cms-sw/cmssw/issues/41356